Rewrite FixDisplayMode

nipkownix opened the pull request:
This should improve the situation in #411.
``FixDisplayMode`` currently works fine, but I was thinking we could still improve it a bit, so this is what I came up with:

- Get rid of the ``CustomRefreshRate`` int. Read the refresh rate from the game's ``config.ini`` instead.
- Provide the option to choose the refresh rate using the game's config menu. To do this, we replace ``D3D_FillDisplayModeVector`` with our own function to make the vector of resolutions. This lets us have a finer control of the list, allowing us to have duplicated resolutions with different refresh rates:
![image](https://user-images.githubusercontent.com/33870163/210075861-58208b83-21e6-4792-92ad-2cd1a8160837.png)
![image](https://user-images.githubusercontent.com/33870163/210075958-9f988554-cb5c-454c-abfe-48a9e638a653.png)
- Hook ``D3D_GetDisplayModeForWidthHeight`` to make it get the display mode with the refresh rate we want, instead of only being able to get 60 Hz refresh rates.
- At startup, validate the display mode in the game's ``config.ini``. Use it if the primary display supports it, otherwise fall back to using the desktop's resolution and refresh rate. (This is and the original ``D3D_GetDisplayModeForWidthHeight`` behavior are what would cause the "black screen" bug originally).
- Hook ``ConfigWriteINI`` to make sure the current resolution is being saved correctly.

An option is also provided to show only the highest refresh rate for each resolution.

This should make the game work with any refresh rate, above or below 60 Hz.

Should probably test it a bit more before merging. But there's one issue: I can't test if this 100% still fixes the original black screen issue, since I was never able to reproduce that in the first place. All my displays have 60 Hz modes. @emoose, maybe we should ping someone here who has the issue? I don't know anyone, though...

Edit: Also decided to update our window hooks a bit. Borderless Window mode can now be toggled at run time, added an option to resize the game window by dragging the window borders, and made the v-sync toggle also work at runtime.
We're also now hooking ``WndProc`` using a slightly better method (in theory...).

TODO: Move the ``GX_DisplayMode`` struct somewhere in the SDK?

emoose commented:
Just got around to trying this, sorry for the delay - I don't really have a monitor higher than 60 neither, but this monitor can be OC'd to 70Hz so thought I'd give that a try.

First thing I tried was running re4_tweaks debug build, game loaded fine, but when I opened the game menu I got an exception I never seen before:
>Exception thrown at 0x0073FFF6 in bio4.exe: 0xC0000005: Access violation reading location 0x1700BFF4.

(addrs are 1.1.0)

Crash seems mostly consistent, another run gave me:
>Exception thrown at 0x0073FFFC in bio4.exe: 0xC0000005: Access violation reading location 0x974B45CC.

Switching to release build lets it open the menu fine though - not really sure what cause of this could be, though I did notice we're using `std::vector<GX_DisplayMode>` in one of our hooks arguments:
>void D3D_FillDisplayModeVector_hook(std::vector<GX_DisplayMode>* vect, bool checkWidthHeightCombinations)

That would mean we're mixing whatever version of `std::vector` that re4_tweaks uses with the games `std::vector`, kinda surprised that works really :p
Looking at std::vector definition it seems pretty simple though, just pointer to start & pointer to end, so guess that's why it works fine, maybe in debug something gets changed there though, I guess we could just create our own `bio4vector` type or something if that's the case.

Unfortunately not sure if I can test much more, seems the game doesn't detect my 70Hz mode, does it maybe only see ones thar are multiple of 30 or something like that?
Don't think I can hide my monitors 60Hz mode neither, so probably can't check if it helps with black-screen issue... maybe worth asking on steam forums to see if anyone there has it?

nipkownix commented:
> Just got around to trying this, sorry for the delay 

No problem, been pretty busy too. Even got a job interview tomorrow :p

> That would mean we're mixing whatever version of std::vector that re4_tweaks uses with the games std::vector, kinda surprised that works really :p

Wasn't too sure about that either, but since it worked.. I thought I had just lucked out :p
I hadn't tested debug builds though. Meh.

>  seems the game doesn't detect my 70Hz mode, does it maybe only see ones that are multiple of 30 or something like that?

Ah, that's odd. Is 60Hz the only mode shown in the menu?
Could you enable ``VerboseLog`` and check which resolutions are logged? IIRC our old method of getting the desktop refresh rate was only querying the first display, not the primary display. ``HMONITOR primaryMonitor = MonitorFromPoint({ 0, 0 }, MONITOR_DEFAULTTOPRIMARY);`` should be getting the actual primary monitor info now, and I checked that this implementation now works fine with my HDMI TV. I can see all types of odd resolutions too (my TV has stuff like 63Hz for some reason).

Do you happen to have another display plugged or something like that? Pretty strange.

emoose commented:
Ah should have posted the list with that, my bad 😅 

```
[02:10:51] [t 41720] +-----------------+-----------------+
[02:10:51] [t 41720] | Display modes                     |
[02:10:51] [t 41720] +-----------------+-----------------+
[02:10:51] [t 41720] | \\.\DISPLAY1: 640  x 480    59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 640  x 480    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 720  x 480    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 720  x 480    59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 720  x 576    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 720  x 576    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 720  x 576    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 800  x 600    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1024 x 768    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1152 x 864    59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1152 x 864    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1176 x 664    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1176 x 664    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1176 x 664    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1176 x 664    59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1176 x 664    59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1176 x 664    59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1280 x 720    59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1280 x 720    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1280 x 720    59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1280 x 720    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1280 x 720    59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1280 x 720    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1280 x 768    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1280 x 768    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1280 x 768    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1280 x 800    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1280 x 800    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1280 x 800    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1280 x 960    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1280 x 960    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1280 x 960    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1280 x 1024   60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1360 x 768    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1360 x 768    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1360 x 768    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1360 x 768    59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1360 x 768    59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1360 x 768    59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1360 x 768    50 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1360 x 768    50 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1360 x 768    50 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1360 x 768    24 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1360 x 768    24 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1360 x 768    24 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1360 x 768    23 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1360 x 768    23 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1360 x 768    23 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1360 x 768    25 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1360 x 768    25 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1360 x 768    25 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1360 x 768    30 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1360 x 768    30 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1360 x 768    30 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1360 x 768    29 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1360 x 768    29 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1360 x 768    29 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1366 x 768    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1366 x 768    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1366 x 768    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1366 x 768    59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1366 x 768    59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1366 x 768    59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1366 x 768    50 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1366 x 768    50 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1366 x 768    50 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1366 x 768    24 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1366 x 768    24 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1366 x 768    24 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1366 x 768    23 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1366 x 768    23 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1366 x 768    23 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1366 x 768    25 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1366 x 768    25 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1366 x 768    25 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1366 x 768    30 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1366 x 768    30 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1366 x 768    30 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1366 x 768    29 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1366 x 768    29 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1366 x 768    29 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 900    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 900    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 900    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 900    59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 900    59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 900    59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 900    50 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 900    50 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 900    50 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 900    24 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 900    24 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 900    24 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 900    23 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 900    23 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 900    23 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 900    25 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 900    25 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 900    25 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 900    30 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 900    30 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 900    30 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 900    29 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 900    29 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 900    29 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 1024   60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 1024   60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 1024   60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 1024   59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 1024   59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 1024   59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 1024   50 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 1024   50 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 1024   50 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 1024   24 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 1024   24 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 1024   24 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 1024   23 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 1024   23 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 1024   23 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 1024   25 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 1024   25 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 1024   25 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 1024   30 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 1024   30 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 1024   30 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 1024   29 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 1024   29 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 1024   29 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1600 x 1200   60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1680 x 1050   60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1680 x 1050   60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1680 x 1050   60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1680 x 1050   59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1680 x 1050   59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1680 x 1050   59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1680 x 1050   50 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1680 x 1050   50 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1680 x 1050   50 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1680 x 1050   24 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1680 x 1050   24 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1680 x 1050   24 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1680 x 1050   23 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1680 x 1050   23 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1680 x 1050   23 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1680 x 1050   25 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1680 x 1050   25 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1680 x 1050   25 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1680 x 1050   30 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1680 x 1050   30 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1680 x 1050   30 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1680 x 1050   29 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1680 x 1050   29 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1680 x 1050   29 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1920 x 1080   60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1920 x 1080   59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1920 x 1080   50 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1920 x 1080   24 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1920 x 1080   23 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1920 x 1080   25 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1920 x 1080   30 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1920 x 1080   29 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1920 x 1200   59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1920 x 1200   60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1920 x 1200   59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1920 x 1200   60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1920 x 1200   59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1920 x 1200   60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1920 x 1440   59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1920 x 1440   60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1920 x 1440   59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1920 x 1440   60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1920 x 1440   59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1920 x 1440   60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1506 x 847    59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1506 x 847    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1506 x 847    59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1506 x 847    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1506 x 847    59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1506 x 847    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1707 x 960    59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1707 x 960    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1707 x 960    59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1707 x 960    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1707 x 960    59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1707 x 960    60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1969 x 1108   59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1969 x 1108   60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1969 x 1108   59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1969 x 1108   60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1969 x 1108   59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 1969 x 1108   60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 2176 x 1224   59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 2176 x 1224   60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 2176 x 1224   59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 2176 x 1224   60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 2176 x 1224   59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 2176 x 1224   60 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 2560 x 1440   59 Hz |
[02:10:51] [t 41720] | \\.\DISPLAY1: 2560 x 1440   60 Hz |
[02:10:51] [t 41720] +-----------------+-----------------+
```

Game shows the 2560x1440 59Hz and 60Hz modes fine, this is with one display connected too. IIRC I had issues trying to get RE4 to see the OC'd 70hz mode in the past as well (think the only way I've got it to run at 70Hz was with game in windowed mode with Windows display set to 70Hz, and the variable frametime stuff enabled)
Also some games do detect it fine but not all, maybe just an issue with whatever method NV uses to create the custom modes... but then again the other PR with the older FixDisplayMode code did seem to recognize 70Hz somehow?

There's a bunch of weird ones there like 2176 x 1224 that my monitor doesn't actually support, think that's something to do with nvidia NIS, at least I see those ones mentioned in NVCP:
![image](https://user-images.githubusercontent.com/1755499/212794914-26feaa91-6c4a-4255-8339-0e77959ff244.png)

Good luck with your job interview too!

emoose commented:
Ah seems to be an NV problem, tried setting 55Hz and it also didn't show, changed to using CRU to add 70Hz mode instead and that got it to work 🎉 
```
[02:39:55] [t 2784] | 2560 x 1440                 60 Hz |
[02:39:55] [t 2784] | 2560 x 1440                 71 Hz |
[02:39:55] [t 2784] +-----------------+-----------------+
```
Was able to select 71Hz in the game settings too, not totally sure if it was running in that mode though, FPS was still capped to 60... will try messing with variableframetime setting now.

nipkownix commented:
> good luck with your job interview too!

Many thanks! Things went well, actually. Hoping to hear from them again soon 😅

> Ah seems to be an NV problem, tried setting 55Hz and it also didn't show, changed to using CRU to add 70Hz mode instead and that got it to work 🎉

Ah, awesome! Should work fine on "real" refresh rates then. Thanks for looking into it!
Rebased on top of current master, shooould be working fine.
Have you tried the window hooks btw? (adding/removing the border in real time).

Also, looked into the std::vector issue, but I'm not sure how to solve that issue. Meh :p

nipkownix commented:
Worked around the debug build crash by using the game's own ``std::vector<D3D_DisplayMode>::push_back`` in our hook. Seems to work fine, but... That's a weird issue. Seems like the game would crash whenever I tried to manipulate the ``*vect``, not matter what I tried. Still now sure what was different with debug builds, but this new method seems to work at least.

emoose commented:
Oh nice solution, seems to work fine for me in debug now. I can't really see what would make debug builds break neither, using games func is probably best idea though.

WindowBorderless seems to work fine too, I did notice it reset the window position though (when RememberWindowPos / X Pos / Y Pos were all unset, I just alt-tabbed out & moved window before changing borderless option)

Also if I do set RememberWindowPos and then move it & toggle borderless, it does seem to use the saved X position at least, but didn't use the saved Y position (you can see that by setting X Pos/Y Pos to 500 and then toggle the option, always seems to use Y = 0)

nipkownix commented:
Ah, thanks for catching those! Latest commit should fix these issues.
